File: ios/Earwyrm/Core/Models/Lyric.swift
import Foundation

struct Lyric: Codable, Identifiable {
    let id: UUID
    let userId: UUID
    let content: String
    let songTitle: String?
    let artistName: String?
    let coverArtUrl: String?
    let albumName: String?
    let isCurrent: Bool?
    let isPublic: Bool?
    let tags: [String]?
    let shareToken: String?
    let canonicalLyricId: UUID?
    let musicbrainzRecordingId: String?
    let musicbrainzReleaseId: String?
    let reactionCount: Int?
    let commentCount: Int?
    let createdAt: Date
    let replacedAt: Date?

    enum CodingKeys: String, CodingKey {
        case id
        case userId = "user_id"
        case content
        case songTitle = "song_title"
        case artistName = "artist_name"
        case coverArtUrl = "cover_art_url"
        case albumName = "album_name"
        case isCurrent = "is_current"
        case isPublic = "is_public"
        case tags
        case shareToken = "share_token"
        case canonicalLyricId = "canonical_lyric_id"
        case musicbrainzRecordingId = "musicbrainz_recording_id"
        case musicbrainzReleaseId = "musicbrainz_release_id"
        case reactionCount = "reaction_count"
        case commentCount = "comment_count"
        case createdAt = "created_at"
        case replacedAt = "replaced_at"
    }
}

File: android/app/src/main/java/com/earwyrm/app/core/model/Lyric.kt
package com.earwyrm.app.core.model

import kotlinx.datetime.Instant
import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable

@Serializable
data class Lyric(
    val id: String, @SerialName("user_id") val userId: String, val content: String,
    @SerialName("song_title") val songTitle: String? = null,
    @SerialName("artist_name") val artistName: String? = null,
    @SerialName("cover_art_url") val coverArtUrl: String? = null,
    @SerialName("album_name") val albumName: String? = null,
    @SerialName("is_current") val isCurrent: Boolean? = null,
    @SerialName("is_public") val isPublic: Boolean? = null,
    val tags: List<String>? = null,
    @SerialName("share_token") val shareToken: String? = null,
    @SerialName("canonical_lyric_id") val canonicalLyricId: String? = null,
    @SerialName("musicbrainz_recording_id") val musicbrainzRecordingId: String? = null,
    @SerialName("musicbrainz_release_id") val musicbrainzReleaseId: String? = null,
    @SerialName("reaction_count") val reactionCount: Int? = null,
    @SerialName("comment_count") val commentCount: Int? = null,
    @SerialName("card_art_url") val cardArtUrl: String? = null,
    @SerialName("created_at") val createdAt: Instant,
    @SerialName("replaced_at") val replacedAt: Instant? = null
)

@Serializable
data class LyricInsert(
    @SerialName("user_id") val userId: String, val content: String,
    @SerialName("song_title") val songTitle: String? = null,
    @SerialName("artist_name") val artistName: String? = null,
    @SerialName("cover_art_url") val coverArtUrl: String? = null,
    @SerialName("card_art_url") val cardArtUrl: String? = null,
    @SerialName("album_name") val albumName: String? = null,
    val tags: List<String>? = null,
    @SerialName("is_public") val isPublic: Boolean = true,
    @SerialName("is_current") val isCurrent: Boolean = true,
    @SerialName("canonical_lyric_id") val canonicalLyricId: String? = null,
    @SerialName("musicbrainz_recording_id") val musicbrainzRecordingId: String? = null,
    @SerialName("musicbrainz_release_id") val musicbrainzReleaseId: String? = null
)

@Serializable
data class LyricUpdate(
    val content: String? = null, @SerialName("song_title") val songTitle: String? = null,
    @SerialName("artist_name") val artistName: String? = null,
    @SerialName("cover_art_url") val coverArtUrl: String? = null,
    @SerialName("album_name") val albumName: String? = null,
    val tags: List<String>? = null,
    @SerialName("is_public") val isPublic: Boolean? = null,
    @SerialName("is_current") val isCurrent: Boolean? = null,
    @SerialName("replaced_at") val replacedAt: Instant? = null,
    @SerialName("canonical_lyric_id") val canonicalLyricId: String? = null,
    @SerialName("musicbrainz_recording_id") val musicbrainzRecordingId: String? = null,
    @SerialName("musicbrainz_release_id") val musicbrainzReleaseId: String? = null
)
